tensorpipe/channel/cuda_basic/channel.cc

```diff
@@ -0,0 +1,358 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <tensorpipe/channel/cuda_basic/channel.h>
+
+#include <algorithm>
+#include <condition_variable>
+#include <list>
+#include <mutex>
+
+#include <tensorpipe/channel/cuda_basic/context_impl.h>
+#include <tensorpipe/channel/error.h>
+#include <tensorpipe/channel/helpers.h>
+#include <tensorpipe/common/callback.h>
+#include <tensorpipe/common/cuda.h>
+#include <tensorpipe/common/cuda_loop.h>
+#include <tensorpipe/common/defs.h>
+#include <tensorpipe/common/error.h>
+#include <tensorpipe/common/error_macros.h>
+#include <tensorpipe/transport/connection.h>
+
+namespace tensorpipe {
+namespace channel {
+namespace cuda_basic {
+
+namespace {
+
+std::shared_ptr<uint8_t> makeCudaPinnedBuffer(size_t length) {
```

Comment:
Nit: any reason for this to be a shared_ptr rather than a unique_ptr? I know that declaring a custom deleter through a lambda is easier than defining a custom struct, but I like the stronger guarantees of unique_ptr. Also, this could go in common/cuda.h since it seems quite reusable.